File: 278_first_bad_version/firstBadVersion.kt
/* The isBadVersion API is defined in the parent class VersionControl.
      fun isBadVersion(version: Int) : Boolean {} */

class Solution: VersionControl() {
    override fun firstBadVersion(n: Int) : Int {
            var l: Long = 1
    var r: Long = n.toLong() + 1
    
    while (l < r) {
        val m = l + (r - l) / 2

        if (isBadVersion(m.toInt()))
            r = m
        else
            l = m + 1
    }

    return l.toInt()
	}
}

File: 278_first_bad_version/first_bad_version.rs
// The API isBadVersion is defined for you.
// isBadVersion(version:i32)-> bool;
// to call it use self.isBadVersion(version)

impl Solution {
    pub fn first_bad_version(&self, n: i32) -> i32 {
		let mut l = 1;
        let mut r = n;

        while l <= r {
            let mut m = l + (r - l) / 2;

            if self.isBadVersion(m) {
                r = m - 1
            } else {
                l = m + 1
            }
        }

        return l
    }
}